### PR TITLE
[Feature]: Add Excel Tool for reading/writing .xlsx/.xlsm files #2675

### LINKED ISSUE #2675: [Feature]: Add Excel Tool for reading/writing .xlsx/.xlsm files
## Problem Statement

Agents working with enterprise data frequently encounter Excel files (.xlsx, .xlsm), which are fundamentally different from CSVs:

1. **Multi-sheet structure** - Excel workbooks contain multiple sheets, each a separate data table. Agents need to discover sheets before reading.
2. **No existing tooling** - The `csv_tool` handles single-table files well, but there's no equivalent for Excel workbooks.
3. **Enterprise ubiquity** - Excel is the default format for business reports, financial data, and cross-team data sharing.

Without Excel support, agents cannot process a significant portion of real-world business data files.

## Proposed Solution

Add an `excel_tool` module following the established `csv_tool` pattern with **4 MCP tools**:

| Tool | Purpose |
|------|---------|
| `excel_list_sheets` | List all sheets with metadata (row counts, columns, hidden status) |
| `excel_read` | Read data from a specific sheet with pagination |
| `excel_info` | Quick file metadata (faster, no content scanning) |
| `excel_write` | Create/modify workbooks with replace/append modes |

**Key design decisions:**
- **Deterministic default**: `sheet=None` reads first sheet (index 0), not the ambiguous "active" sheet
- **Discovery-first workflow**: `excel_read` response always includes `available_sheets` so agents know what else exists
- **Data type preservation**: Dates → ISO strings, formulas → calculated values, booleans preserved
- **Optional dependency**: Uses `openpyxl` via `pip install aden-tools[excel]` (doesn't bloat base install)
                                                                                                                      
  

## Alternatives Considered

1. **Extend csv_tool** - Rejected because Excel's multi-sheet nature requires fundamentally different API design (sheet discovery, sheet selection)

2. **Use pandas read_excel** - Rejected because pandas is a heavy dependency; openpyxl is lightweight and purpose-built

3. **Support .xls (legacy format)** - Deferred; openpyxl only supports modern .xlsx/.xlsm. Legacy .xls would require additional `xlrd` dependency

## Additional Context

**Alignment with ROADMAP.md:**
- Extends "Essential Tools" section following established patterns
- Follows "File Use Tool Kit" architecture (sandbox security via `get_secure_path`)
- Complements existing `csv_tool` for complete tabular data coverage

**Implementation ready:**
- Full implementation complete with 35 tests
- Follows `BUILDING_TOOLS.md` checklist
- All tests pass, linting clean

## Implementation Ideas

```
tools/src/aden_tools/tools/excel_tool/
├── __init__.py           # Export register_tools
├── excel_tool.py         # 4 tools (~350 lines)
└── README.md             # Documentation

tools/tests/tools/test_excel_tool.py  # 35 tests
```

**Test coverage includes:**
- Sheet discovery (single, multi, hidden sheets)
- Read operations (by name, by index, pagination, empty sheets)
- Data types (dates, booleans, floats, formulas)
- Edge cases (duplicate columns, special characters in sheet names)
- Write operations (create, replace, append)
- Error handling (file not found, wrong extension, path traversal)
- Graceful degradation when openpyxl not installed

### PR DESCRIPTION
**Pull request by Amdev-5**
## Description

Adds comprehensive Excel spreadsheet tools to the Aden tools library, enabling agents to read, write, query, and search Excel files 

## Type of Change

- [x] New feature (non-breaking change that adds functionality)

## Related Issues

Fixes #2675 

### New [excel_tool](cci:7://file:///Users/siketyson/Desktop/aden/tools/src/aden_tools/tools/excel_tool:0:0-0:0) Module

Added 7 tool functions for complete Excel file manipulation:

| Tool | Description |
|------|-------------|
| `excel_read` | Read Excel files with pagination support (limit/offset) |
| `excel_write` | Create new Excel files with custom sheet names |
| `excel_append` | Append rows to existing sheets |
| `excel_info` | Get file metadata (sheets, columns, row counts) |
| `excel_sheet_list` | List all sheet names in a workbook |
| `excel_sql` | Query Excel data with SQL via DuckDB (including multi-sheet joins!) |
| `excel_search` | Search values across sheets with `contains`/`exact`/`starts_with` matching |

### Other Changes

- Added `openpyxl` as optional dependency (`pip install aden-tools[excel]`)
- Updated [__init__.py](cci:7://file:///Users/siketyson/Desktop/aden/tools/src/aden_tools/tools/__init__.py:0:0-0:0) to register Excel tools
- Added Excel tools reference to `building-agents-construction/SKILL.md`

## Testing

- ✅ **56 new unit tests** covering all Excel functions
- ✅ Lint passes (`ruff check . && ruff format .`)
- ✅ Manual testing performed

```bash
# Run tests
pytest tools/tests/tools/test_excel_tool.py -v

# Example output
56 passed in 2.34s

### CODE REVIEW

**Comment by Amdev-5** [comment]
@Hundao   can you review the PR

**Comment by Amdev-5** [comment]
@Hundao @RichardTang-Aden can we merge this

File: tools/src/aden_tools/tools/excel_tool/excel_tool.py [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+                    headers = []
+                    first_row = next(ws.iter_rows(min_row=1, max_row=1, values_only=True), None)
+                    if first_row:
+                        headers = [
+                            str(c) if c is not None else f"Column_{i + 1}"
+                            for i, c in enumerate(first_row)
+                        ]
+
+                    # Search all cells
+                    for row_idx, row in enumerate(ws.iter_rows(values_only=True), start=1):
```
Comment (Hundao) [inline tools/src/aden_tools/tools/excel_tool/excel_tool.py]: Iterates from row 1 which includes the header row. Searching for "name" would match the column header itself. Should start from row 2, or flag header matches separately.

File: tools/src/aden_tools/tools/excel_tool/excel_tool.py [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+                    else:
+                        # Empty table
+                        cols_sql = ", ".join(f'"{h}" VARCHAR' for h in headers)
+                        con.execute(f'CREATE TABLE "{table_name}" ({cols_sql})')
+
+                    # Create 'data' alias for target sheet
+                    if sheet_name == target_sheet:
+                        con.execute(f'CREATE VIEW data AS SELECT * FROM "{table_name}"')
+
+                wb.close()
```
Comment (Hundao) [inline tools/src/aden_tools/tools/excel_tool/excel_tool.py:622]: The workbook is closed here but wb.sheetnames is accessed later in the return dict(available_sheets). It works because openpyxl caches the attribute, but it's fragile. Save it to a variable before closing.

**Review by Hundao — CHANGES_REQUESTED** [review]

**Comment by Amdev-5** [comment]
Addressed all requested feedback on this PR and revalidated end-to-end.

  Changes made for review comments:
  1) `excel_search` now skips header row (search starts from row 2).
  2) `excel_sql` now snapshots `sheetnames` before workbook close and uses that snapshot in response metadata.

  Validation run:
  - ruff check: pass
  - ruff format --check: pass
  - excel tool tests: 61 passed
  - full tools tests: 895 passed, 18 skipped
  - core tests: 722 passed, 54 skipped

  PR is now on commit `94e9b93`.
  Could you please take another look? Thanks.

**Comment by Amdev-5** [comment]
@Hundao
